picard/mbjson.py

```diff
@@ -192,46 +193,52 @@ def _translate_artist_node(node):
-                elif alias['locale'].split("_")[0] == lang:
-                    score = 0.4
-                else:
-                    continue
-                parts.append((score, 5))
-                if alias["type"] == "Artist name":
-                    score = 0.8
-                elif alias["type"] == "Legal Name":
-                    score = 0.5
-                else:
-                    # as 2014/09/19, only Artist or Legal names should have the
-                    # Primary flag
-                    score = 0.0
-                parts.append((score, 5))
-                comb = linear_combination_of_weights(parts)
-                if comb > result[0]:
-                    result = (comb, (alias['name'], alias["sort-name"]))
-            transl, translsort = result[1]
-        if not transl:
-            translsort = node['sort-name']
-            transl = translate_from_sortname(node['name'] or "", translsort)
+        if config.setting['translate_artist_names_script_exception']:
+            threshhold = config.setting["artist_script_exception_weighting"] / 100
+            detected_scripts = list_script_weighted(node["name"], threshhold)
+            for script_id in config.setting["artist_script_exceptions"]:
+                if script_id in detected_scripts:
+                    return node['name'], node['sort-name']
+
+        for locale in config.setting["artist_locales"]:
+            lang = locale.split("_")[0]
+
+            if "aliases" in node:
+                result = (-1, (None, None))
+                for alias in node['aliases']:
+                    if not alias["primary"]:
+                        continue
+                    if "locale" not in alias:
+                        continue
+                    parts = []
+                    if alias['locale'] == locale:
+                        score = 0.8
+                    elif alias['locale'] == lang:
+                        score = 0.6
+                    elif alias['locale'].split("_")[0] == lang:
+                        score = 0.4
+                    else:
+                        continue
+                    parts.append((score, 5))
+                    if alias["type"] == "Artist name":
+                        score = 0.8
+                    elif alias["type"] == "Legal Name":
+                        score = 0.5
+                    else:
+                        # as 2014/09/19, only Artist or Legal names should have the
+                        # Primary flag
+                        score = 0.0
+                    parts.append((score, 5))
+                    comb = linear_combination_of_weights(parts)
+                    if comb > result[0]:
+                        result = (comb, (alias['name'], alias["sort-name"]))
+                transl, translsort = result[1]
+                if transl:
+                    # Matching locale found.  Skip further processing.
+                    return (transl, translsort)
+            if not transl:
```

Comment:
> I think this part of the code needs a rewrite.

I agree.  In fact, I thought about this overnight (couldn't sleep) and I want to change the logic a bit so that it does a first pass through the selected locales list looking for an exact match on the full locale, and if there is no match then do another pass looking for a match on the root locales from the items in the selected locales list.  That should provide the user more control if they specify multiple sub-locales for the same root locale.  To make this a bit more efficient, I plan to do a single pass through the aliases and create a dictionary of available aliases based on the best score for each locale found.  The selected alias checks would then be done against this dictionary rather than parsing the aliases list multiple times and recalculating the scores for the items each time.